### PR TITLE
ONE-191: Added Constants and Models from ClientSDK

### PR DESCRIPTION
**Pull request by stevenhaching**
See Also: https://crucible.waterqualitytools.com/cru/CR-FF-8786

### CODE REVIEW

**Comment by stevenhaching** [comment]
Made mistake and set assignees instead of reviewers

File: src/ONE.Models.CSharp/Enums/EnumLogLevel.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    public enum EnumLogLevel
+    {
+        Unknown = 0,
+        Fatal = 1,
+        Error = 2,
+        Warn = 3,
+        Info = 4,
+        Debug = 5,
+        Trace = 6
+    }
```
Comment (MartinBenn-AI) [inline src/ONE.Models.CSharp/Enums/EnumLogLevel.cs:18]: Doesn't really matter, but you can omit the numbers after the 1st one and it will do the counting for you:
```suggestion
    public enum EnumLogLevel
    {
        Unknown = 0,
        Fatal,
        Error,
        Warn,
        Info,
        Debug,
        Trace
    }
```

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/WorksheetView.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        public string name { get; set; }
+        public bool? isOwner { get; set; }
+        public EnumWorksheet worksheetType { get; set; }
+
+        public List<uint> columnNumbers { get; set; }
+        public Dictionary<uint, ColumnConfig> columns { get; set; }
+        public string createdOn { get; set; }
+        public bool? canEdit { get; set; }
+        public bool? canDelete { get; set; }
+        public List<dynamic> headers { get; set; } // (ColumnHeader | GroupHeader)
```
Comment (MartinBenn-AI) [inline src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/WorksheetView.cs:17]: I'm not sure how this is used but using dynamic as a workaround for the lack of discriminated union in C# seems a bit heavy handed since you've no way of knowing what data structure you're going to be given when looking at an instance of this class and makes unit testing, serializing / deserializing a bit harder.

Might be clearer to pull these out into two different lists (columns, groups) that way the consumer doesn't need to guess at what it is dealing with and someone can't jam the wrong thing in. Otherwise, an interface could be used to at least restrict the types to those that are expected in this collection.

File: src/ONE.Models.CSharp/UI/Attribute.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
+using ONE.Models.CSharp.Enums;
+
+namespace ONE.Models.CSharp.UI
+{
+    public class Attribute
```
Comment (MartinBenn-AI) [inline src/ONE.Models.CSharp/UI/Attribute.cs:6]: Not really clear to me what this class is supposed to represent. It looks like this is supposed to be the UI Definition based on the properties, but with that definition property it isn't clear.

**Review by MartinBenn-AI — COMMENTED** [review]
Not sure on the context for these existing model so most my comments might not be applicable.

File: src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ONE.Models.CSharp.Common.Configuration
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs:5]: Having Common.Configuration in the namespace indicates that these are models used by the Common.Configuration MSApp which is not the case.  I might suggest a One.Models.CSharp.Constants namespace and then it can include classes like ConfigurationTypes, ActivityTypes, Roles, etc.

File: src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            public const string Id = "9a4a09ca-cf6c-40c0-bc21-1b8f5d306e0a";
+            public static readonly Guid Guid = new Guid(Id);
+        }
+        public static class Graphs
+        {
+            public const string Id = "24ecb22b-aee3-4bac-a88d-9e10402363ba";
+            public static readonly Guid Guid = new Guid(Id);
+        }
+        public static class Worksheets
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs:19]: Minor issue but this is technically a WorksheetView

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections.Generic;
+
+namespace ONE.Models.CSharp.Models.Common.Configuration.Graphs
+{
+    public class GraphConfigurationData
+    {
+        public string name { get; set; }
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs:7]: Property names in here should be capitalized

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        public int defaultDateRange { get; set; }
+        public List<int> columns { get; set; }
+        public List<Series> series { get; set; }
+
+        public class Series
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs:12]: I think there is a Series protobuf object is this related to that in anyway?  If not perhaps it should be renamed?

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphAxisMode.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿namespace ONE.Models.CSharp.Models.Common.Configuration.Graphs
+{
+    public enum IGraphAxisMode
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphAxisMode.cs:3]: This enum should go in the Enums folder

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphConfiguration.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            "twinId": 231138,
+            "color": "#520408",
+            "showLimits": false,
+            "type": "bar",
+            "axis": "1"
+        }
+    ],
+    "axisMode": "2"
+     */
+    public interface IGraphConfiguration
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphConfiguration.cs:32]: Might be good to have interfaces in their own folder.  Maybe the top level folders could be something like

- Constants
- Enums
- Interfaces
- ... Haven't figured out the rest yet

Also this GraphConfiguration interface is not implemented by the GraphConfigurationData class which seems odd.

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs [hunk 10]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace ONE.Models.CSharp.Models.Common.Configuration.Spreadsheet
```
[the remark this answers is not in the document]
Comment (AQI-DanG) [reply src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs:3]: Additionally Common.Configuration.Spreadsheet is very confusing as that encompasses to different MSApps in two different domains.

Not sure where to put these models really but this organization does not feel great.  Currently all of the protobuf models are in the One.Models.CSharp namespace.  Perhaps these models could also reside there or perhaps they should be in One.Models.CSharp.SDK or One.Models.CSharp.Extras or something similar.

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/GroupHeader.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections.Generic;
+using ONE.Models.CSharp.Enums;
+
+namespace ONE.Models.CSharp.Models.Common.Configuration.Spreadsheet
+{
+    public class GroupHeader
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/GroupHeader.cs:6]: It seems like most of these classes are basically an effort to strongly type json data.  Perhaps the namespace/naming convention should reflect that.
Additionally it has been suggested that this sort of thing should be added to the protobuf models so that we don't have everyone writing their own POCO to translate the same json data.

File: src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs [hunk 12]
```diff
@@ -0,0 +1,50 @@
+﻿using System;
+
+namespace ONE.Models.CSharp.Enterprise.Twin
+{
+    public static class InstrumentCategoryConstants
+    {
+        public const int Id = 3;
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs:7]: This seems awkward and out of place.

File: src/ONE.Models.CSharp/Enterprise/Twin/OrganizationCategoryConstants.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace ONE.Models.CSharp.Enterprise.Twin
+{
+    public static class OrganizationCategoryConstants
+    {
+        public const int Id = 1;
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Enterprise/Twin/OrganizationCategoryConstants.cs:7]: Ah, now I remember this is the Id of the category of twins in this class.

File: src/ONE.Models.CSharp/Enterprise/Twin/OrganizationCategoryConstants.cs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+        public static class TenantType
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Enterprise/Twin/OrganizationCategoryConstants.cs:9]: I might suggest this gets moved to a Constants.TwinTypes namespace

File: src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ONE.Models.CSharp.Enterprise.Twin.Wizard
+{
+    public class Location
+    {
+        // Is this a default in the wizard
+        public bool d { get; set; }
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs:10]: These names are not good, is this data being deserialized from something that has equally poor naming?

File: src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs [hunk 16]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ONE.Models.CSharp.Enterprise.Twin.Wizard
+{
+    public class Location
```
Comment (AQI-DanG) [inline src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs:7]: Isn't there already a location protobuf?

**Review by AQI-DanG — CHANGES_REQUESTED** [review]
I think this effort presents an opportunity to clean up a lot of code to make it more production ready.  I would hate to just copy/paste the same thing in the name of expediency.
My suggestion for top level organization would be something like the following.

1. Constants - for classes that are basically documentation of required guids and other things.  TwinTypes could be a folder under this.
2. Enums - All enums can go here
3. Interfaces - All interfaces can go here
4. Implementations - Implementations of the above interfaces can go here
5. Wizard ?? Seems like there are some unusal classes specific to this, maybe optional
6. Wims - Not sure if there were models related to WIMS in these changes or not

Anything left that is a simple model used in an MSApp or the SDK that does not fit in the above categories should probably be moved into the protobuf library in a .proto file.

File: src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        public static class DataSourceDriverType
+        {
+            public const string RefId = "3209ac8f-640b-4f68-a74b-08b670187182";
+            public static readonly Guid Guid = new Guid(RefId);
+            public static class DataSourceDriverTest
+            {
+                public const string RefId = "cafd1356-7e06-44d9-9f28-3a897fbc4cf8";
+                public static readonly Guid Guid = new Guid(RefId);
+            }
+            public static class DataSourceDriverOpenWeatherMap
```
Comment (yuinlin) [inline src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs:37]: we deleted this driver in https://github.com/AquaticInformatics/aquarius-one-ingest/pull/37  not sure if it means this class can/should be removed

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphAxisMode.cs (hunk 8 above)
Comment (yuinlin) [inline src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphAxisMode.cs:3]: the name tricked me into thinking this was an interface.  but perhaps too late to change since this was the name in the sdk

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnHeader.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
+﻿using ONE.Models.CSharp.Enums;
+
+namespace ONE.Models.CSharp.Models.Common.Configuration.Spreadsheet
+{
+    public class ColumnHeader
+    {
+        public uint id { get; set; }
+        public EnumHeaderType HeaderType
```
[the remark this answers is not in the document]
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnHeader.cs:8]: In some of these models, the case of the JSON is being matched.  I.e. However the JSON was created by the third party, in this case agGrid, is being matched.

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs:3]: The purpose of this is to have a common model for the deserialization of spreadsheet configurations.
@AQI-DanG : how would you suggest namespacing configuration models that apply to different domains?

File: src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs (hunk 4 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs:5]: I think we just need to agree on a standard.  I just refactored it from what you are suggesting to how it is currently implemented through a separate conversation.

File: src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs (hunk 17 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs:37]: I would be reticent to revoke the GUID assigned.  This could be a little working ahead.

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnHeader.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (AQI-DanG) [reply src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnHeader.cs:8]: This seems to lean toward a category of models that are strictly for deserializing third party provided json strings. Perhaps a JsonTypes namespace?

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (AQI-DanG) [reply src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs:3]: This is also a json deserialization.  I would suggest either the above mentioned Json or JsonTypes namespace or add it to the protobuf library.  The term configuration is already pretty loaded so while applicable it might be best to steer clear of it if we can.

File: src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs (hunk 4 above)
Comment (AQI-DanG) [reply src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs:5]: Perhaps an alignment meeting is in order.

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/ColumnConfig.cs]: Moved to Imposed.

File: src/ONE.Models.CSharp/Enterprise/Twin/Wizard/PropertyBag.cs [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class PropertyBag
+    {
+        //Twin Types to be used
+        public string ty { get; set; }
+        // List of Suggested Locations for this Digital Twin Sub Type
+        public List<Location> ls { get; set; }
+        // List of Suggested Parameters
+        public List<Parameter> ps { get; set; }
+        // List of Suggested Computations
```
[the remark this answers is not in the document]
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Enterprise/Twin/Wizard/PropertyBag.cs]: Removed Comment

File: src/ONE.Models.CSharp/UI/I18NKeyTextReference.cs [hunk 20]
```diff
@@ -0,0 +1,7 @@
+﻿namespace ONE.Models.CSharp.UI
+{
+    public class I18NKeyTextReference
+    {
+        public string i18nKey { get; set; }
+    }
```
[the remark this answers is not in the document]
Comment (stevenhaching) [reply src/ONE.Models.CSharp/UI/I18NKeyTextReference.cs]: Changed.

File: src/ONE.Models.CSharp/Enums/EnumLogLevel.cs (hunk 1 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Enums/EnumLogLevel.cs:18]: Noted

File: src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs (hunk 4 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs]: Moved to Constants

File: src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs (hunk 5 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/ConfigurationTypeConstants.cs]: Renamed.

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs (hunk 7 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs]: Renamed to avoid confusion.

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphAxisMode.cs (hunk 8 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/Graphs/IGraphAxisMode.cs]: Fixed

File: src/ONE.Models.CSharp/Enterprise/Twin/OrganizationCategoryConstants.cs (hunk 14 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Enterprise/Twin/OrganizationCategoryConstants.cs]: Moved and renamed.

File: src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs (hunk 15 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs]: renamed.

File: src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs (hunk 16 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Enterprise/Twin/Wizard/Location.cs]: Renamed to avoid confusion.

File: src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs (hunk 17 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Enterprise/Twin/InstrumentCategoryConstants.cs]: Noted

File: src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs (hunk 6 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/Graphs/GraphConfigurationData.cs]: Moved to Imposed

File: src/ONE.Models.CSharp/UI/Attribute.cs (hunk 3 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/UI/Attribute.cs]: Renamed

File: src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/WorksheetView.cs (hunk 2 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/Common/Configuration/Spreadsheet/WorksheetView.cs]: Moved to Imposed

File: src/ONE.Models.CSharp/ONE.Models.CSharp.csproj [hunk 21]
```diff
@@ -27,4 +27,8 @@
     <PackageReference Include="Google.Protobuf" Version="3.19.1" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Imposed\AAA\" />
+  </ItemGroup>
```
Comment (MartinBenn-AI) [inline src/ONE.Models.CSharp/ONE.Models.CSharp.csproj:32]: Should be able to remove this
```suggestion
```

**Review by MartinBenn-AI — APPROVED** [review]
Seems fine to me, could clean up that unused folder reference in the .csproj though.

**Review by AQI-DanG — APPROVED** [review]
I like the new organization

File: src/ONE.Models.CSharp/ONE.Models.CSharp.csproj (hunk 21 above)
Comment (stevenhaching) [reply src/ONE.Models.CSharp/ONE.Models.CSharp.csproj]: Good catch. Done
